scaaml/capture/scope/lecroy/lecroy_communication.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Communication with the LeCroy oscilloscope. Either using pyvisa or socket.
+"""
+
+from abc import ABC, abstractmethod
+
+import pyvisa
+
+from scaaml.capture.scope.lecroy.lecroy_waveform import LecroyWaveform
+
+
+class LeCroyCommunication(ABC):
+    """A class for communication with the LeCroy oscilloscope.
+    """
+
+    def __init__(self, ip_address: str, timeout: float = 5.0):
+        """Connect to the LeCroy oscilloscope.
+
+        Args:
+          ip_address (str): IP address (or hostname) of the oscilloscope.
+          timeout (float): Number of seconds before communication timeout.
+        """
+        self._ip_address = ip_address
+        self._timeout = timeout
+
+    @abstractmethod
+    def connect(self):
+        """Connect to the scope."""
+        return self
+
+    @abstractmethod
+    def close(self) -> None:
+        """Close the scope connection"""
+
+    @abstractmethod
+    def write(self, message: str) -> None:
+        """Write a message to the oscilloscope.
+        """
+
+    @abstractmethod
+    def query(self, message: str) -> str:
+        """Query the oscilloscope.
+        """
+
+    @abstractmethod
+    def get_waveform(self) -> LecroyWaveform:
+        """Get a LecroyWaveform object representing a single waveform.
+        """
+
+    @abstractmethod
+    def query_binary_values(self,
+                            message: str,
+                            datatype="B",
+                            container=None) -> bytearray:
+        """Query binary data."""
+
+
+class LeCroyCommunicationError(Exception):
+    """Custom exception to deal with various possible exceptions."""
+
+
+def make_custom_exception(func):
+    """Decorator to wrap specific exception into a custom one."""
+
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception as error:
+            raise LeCroyCommunicationError("Communication error") from error
+
+    return wrapper
+
+
+class LeCroyCommunicationLXI(LeCroyCommunication):
```

Comment:
Maybe this class should be named `LeCroyCommunicationVisa` because if the connection string that is passed to the pyvisa resource manager looks like `TCPIP::<ip or host>::INSTR` then it is going to use LXI protocol over TCP but if the connection string looks like `USBTMC::<0xUSB VID>::<0xUSB PID>::<SERIAL NUMBER>::INSTR` it's going to use USB TMC protocol over usb.

Reply:
We don't use USB so let's push code refactoring when (and if) we would need it.